Show VM migrations on a bar chart

rszwajko opened the pull request:
Key points:
1. use stack chart to preserve space 
2. use resize observer to adjust the chart size dynamically

TODO:
1. confirm stack chart as preferred option
2. create toggle to switch between supported time frames (1 month or 24h)
3. create custom hook to get data points for the chart - the points needs to be sorted, `x` part need to be human readable
4. fix problems with overview page layout - double scrollbar is visible after the migration card got bigger


![image](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/4f645218-feb7-4984-afb7-4b27307b4cb7)

yaacov commented:
> confirm stack chart as preferred option

I like side be side (https://www.patternfly.org/charts/bar-chart#basic-with-right-aligned-legend)

@sgratch @ahadas can you way your opinions ?

p.s.
we can always change the chart type as needed ( other type of chart ? https://www.patternfly.org/charts/scatter-chart#basic )

yaacov commented:
> fix problems with overview page layout

we can make it a separate card if it helps ...

yaacov commented:
note:
the chart view is a new view, it does not replace the table view, it's in addition to the table view.

rszwajko commented:
> > confirm stack chart as preferred option
> 
> I like side be side (https://www.patternfly.org/charts/bar-chart#basic-with-right-aligned-legend)
> 
> @sgratch @ahadas can you way your opinions ?
> 
 
The problem is limited width. The card needs to be narrow to fit into the layout. With full 30 data points and a legend this will be hard to achieve. Actually it would be better [to reverse the chart](https://www.patternfly.org/charts/bar-chart/#multi-color-ordered-with-bottom-left-aligned-legend) - the page layout works better with data growing vertically (like tables)

yaacov commented:
in the 7-day view we should not have a problem it's 7 data points

in 24 hours view - we have 24 items, but we can do buckets of 6 hours, and then it's only 4 data items ...

yaacov commented:
> create toggle to switch between supported time frames (1 month or 24h)

the issue is 7 days and 24 hours not 1 month and 24 hours

sjd78 commented:
Hey guys, don't forget that the current patternfly.org site is running PF5 components now!  I believe that `@patternfly/react-charts@^7` is tied to PF5.  It think it is still all victory charts underneath, but the PF styles around things are for sure going to be different.

A quick look at console on main show it is still running PF4 (with `"@patternfly/react-charts": "6.94.19",`) -- https://github.com/openshift/console/blob/master/frontend/package.json

Probably a good thing to be mindful of the release selected on the patternfly.org page!

rszwajko commented:
Changes in the recent force push:
1. removed resize observer - the legend and number of data points will not depend on the size of the container - Victory charts follow responsive design so the charts scale automatically
2. moved charts to a new card - not sure how to name it - for now it's `Last Migrations`
3. optimized layout for 2 modes: last 7 days (bucket size: 1 day) and last 24h (bucket size: 4h ) 
4. add [toggle group](http://v4-archive.patternfly.org/v4/components/toggle-group) to switch between supported modes - since we have only 2 of them a dropdown seemed less user friendly to me

TODO:
1. data hook

### 24h view 

![Screenshot from 2023-08-23 11-53-37](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/3760be4e-3d7c-4fa4-a18e-e1d0c98bb5db)

### 7 day view

![Screenshot from 2023-08-23 11-53-27](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/f7caaef5-e02a-4ff3-9a7a-49822401a191)

yaacov commented:
Nice :+1:

a. in the issue it say use kebab for the dropdown to choose time span, please use kebab
b. for now the chart card should be below the table chart (https://github.com/kubev2v/forklift-console-plugin/pull/702#issuecomment-1682489582)
c. please use vertical chart with 3 separate bars for started/failed/finished (https://github.com/kubev2v/forklift-console-plugin/pull/702#issuecomment-1682489582)

rszwajko commented:
Changes in the recent force push:
1. improve resizing of the Overview page by introducing break points:
    2.  2xl (the chart and settings are in the second column), 
    3.  xl (only settings are in the second column)
    4. lg (only one column for all cards)
2. data hook - used standard `useK8sWatchResource` combined with transformation  logic (filtering and formatting depends on the period chosen by the user)

@yaacov 
> a. in the issue it say use kebab for the dropdown to choose time span, please use kebab
>
Done. However the kebab toggle seems to small here - see the screenshots below.

> b. for now the chart card should be below the table chart (https://github.com/kubev2v/forklift-console-plugin/pull/702#issuecomment-1682489582)
>

Done.

> c. please use vertical chart with 3 separate bars for started/failed/finished (https://github.com/kubev2v/forklift-console-plugin/pull/702#issuecomment-1682489582)
>

Done. Looks better if  there are many migrations on the cluster(i.e. generated mock data on the the recording below) but with small data sets it looks empty (i.e. mock data on screenshots below)

![Screenshot from 2023-08-25 20-41-55](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/3340cee0-7520-4b58-8799-22360ed1e4e5)
![Screenshot from 2023-08-25 20-41-46](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/f608b75a-19d4-4049-9d65-7c5115cca25a)
[Screencast from 08-25-2023 06:51:19 PM.webm](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/2ccf0ace-6c43-47b8-8729-8af23e3fa8f7)

yaacov commented on packages/forklift-console-plugin/src/modules/Overview/views/overview/tabs/Details/cards/ChartsCard.tsx:
```diff
@@ -0,0 +1,160 @@
+      <CardBody className="forklift-status-migration">
+        <Flex justifyContent={{ default: 'justifyContentSpaceBetween' }}>
+          <FlexItem align={{ default: 'alignRight' }}>
+            <Dropdown
```
can you use card actions ?
https://www.patternfly.org/2023.02/components/card#header-images-and-actions

yaacov commented on packages/forklift-console-plugin/src/modules/Overview/views/overview/tabs/Details/cards/ChartsCard.tsx:
```diff
@@ -0,0 +1,160 @@
+const toFinished = (m: V1beta1Migration): string => m.status.completed;
+const hasTimestamp = (timestamp: string) => timestamp && DateTime.fromISO(timestamp).isValid;
+const toDateTime = (timestamp: string): DateTime => DateTime.fromISO(timestamp);
+const isLastWeek = (date: DateTime) => date.diffNow('days').get('days') <= 7;
```
using the word "week" is problematic, lets use "last7Days" and "last24Hours" or some other phrase to be more clear we don't mean weeks and days but, N days and M hours.

a. we show last 7 days, may not be a complete one week, can be some of this week and some of another
b. we may want to change the time to 10 days instead of 7 ...

yaacov commented:
Notes:
a. Need to always show the full time frame, for example, 7 days, need to show 7 days, days with 0 may show 0 but the need to be in the graph,
for example in this image we see only 3 days, we should see all 7:

![screenshot-localhost_9000-2023 08 30-18_04_23](https://github.com/kubev2v/forklift-console-plugin/assets/2181522/d2fa5e9c-6c21-49e0-97a0-a788ed88692b)

b. time format should be a "usual" time format, it should not be a number ( again issue a. need all the time-slots feeled)
for example we see 06, we should see 06:00-07:00 for the timeslot between 6 and 7:

![screenshot-localhost_9000-2023 08 30-18_03_48(1)](https://github.com/kubev2v/forklift-console-plugin/assets/2181522/16c990bb-8e8c-4364-889b-7e9409862824)

c. blue as the color for "failed" , hmm, I think red is more appropriate, in the table we use blue for started, green for finish successfully, and red for failed, we should use the same color scheme

yaacov commented:
Note III:
the order of the columns should be like in the table started | finished | failed
in the images, started is the last column instead of first

rszwajko commented:
@yaacov 
>  on my screen the spacings are not the same, the spacing between all cards should be the same:
>

Most likely we need another break point for the biggest screens - however it will be difficult to implement without testing on such screen.

yaacov commented:
> Switched to interval based labels with hour granularity i.e. 14 - 18 which translates to 14:00 - 18:59 with minute granularity. See Luxon's helpers endOf and startOf

please use time format of HH:MM as suggested above,
so the labels can look like 14:00-16:00 or 14:00-15:59 but not "17" and "13-17" like in the image you attached

> finished  !=  finished successfully - currently we count all migrations with status.completed regardless of the cause

if you look at the table you will see it say "successful" so we need to show the migrations that where finished successfully.
if you think we currently show the wrong value, it should be tracked in an issue and fixed.

> PF4 recommends the same color family for the nested charts - https://www.patternfly.org/charts/colors-for-charts#chart-with-nested-grouped-variables

PF suggest using the PatternFly blue colors, but not when the value has meaning of "Danger" and "Success", in this cases "Status and state colors" it suggest red and green colors.
in our cases failed migrations fall into the category of "danger" and successful migration fall into the category of "success"

a. please use "successful" string for the finished migrations, and make sure it actually show the number of successful migrations (fix may be in a separate PR)
b. please update the colors to follow PF  "Status and state colors" color scheme
 
Ref: https://www.patternfly.org/design-foundations/colors#patternfly-blue
Ref: https://www.patternfly.org/design-foundations/colors#status-and-state-colors

yaacov commented:
> Most likely we need another break point for the biggest screens - however it will be difficult to implement without testing on such screen.

best way to start is removing all the size modifiers and use clean PF GridItem with span and rowspan for better maintainability and readability, it will also have better results visually (no strange gaps)

rszwajko commented:
![Screenshot from 2023-09-04 22-28-33](https://github.com/kubev2v/forklift-console-plugin/assets/64194103/3e047660-6fc4-4f3b-be06-d163e7ca758f)

yaacov commented on packages/forklift-console-plugin/src/modules/Overview/views/overview/tabs/Details/cards/ChartsCard.tsx:
```diff
@@ -0,0 +1,196 @@
+
+export const ChartsCard: React.FC<MigrationsCardProps> = () => {
+  const { t } = useForkliftTranslation();
+  const [isDropdownOpened, setIsDropdownOpened] = useState(false);
```
looks like a good use for useToggle

https://github.com/kubev2v/forklift-console-plugin/blob/1c71a49b09f6ad0d1783dfd8d703056a3ada0516/packages/forklift-console-plugin/src/modules/Providers/hooks/useToggle.ts

rszwajko replied:
the proposed implementation is:
```
const [value, setIsOpen] = useState(initialValue);
const toggle = () => { setIsOpen((v) => !v);  };
```
Which does not toggle the internal state but negates provided parameter.
My understanding of toggle is the following:
```
const [value, setIsOpen] = useState(initialValue);
const toggle = () => { setIsOpen(() => !value);  };
```

yaacov commented on packages/forklift-console-plugin/src/modules/Overview/views/overview/tabs/Details/cards/ChartsCard.tsx:
```diff
@@ -0,0 +1,196 @@
+  const { t } = useForkliftTranslation();
+  const [isDropdownOpened, setIsDropdownOpened] = useState(false);
+  const onToggle = () => setIsDropdownOpened(!isDropdownOpened);
+  const [isDaysViewSelected, setIsDaysViewSelected] = useState(true);
```
the selection should be between sets of ranges and bucket sizes:

for example if we have one option:
``` js
export const TIME_RANGE_OPTIONS = [
  {
    label: 'Last 24 hours',
    startAtSecondsAgo: 86400,
    bucketSizeInSeconds: 3600,
  },
];
```

rszwajko replied:
1. `bucketSizeInSeconds` does not scale well beyond hours - check https://moment.github.io/luxon/#/math
2.  we can provide a generic implementation but it will be rather complex - pleas describe what use cases you would like to cover

yaacov replied:
> pleas describe what use cases you would like to cover

I can do a folowup that will make the code more general and readable.

@rszwajko we can leave the code as is for this PR, and I'll do a folowup, that will make the code more general, WDYT ?

yaacov commented on packages/forklift-console-plugin/src/modules/Overview/views/overview/tabs/Details/cards/ChartsCard.tsx:
```diff
@@ -0,0 +1,196 @@
+  const [isDropdownOpened, setIsDropdownOpened] = useState(false);
+  const onToggle = () => setIsDropdownOpened(!isDropdownOpened);
+  const [isDaysViewSelected, setIsDaysViewSelected] = useState(true);
+  const [allMigrations] = useK8sWatchResource<V1beta1Migration[]>({
```
```suggestion
  const [migrations] = useK8sWatchResource<V1beta1Migration[]>({
```

the data for the charts will need different name

yaacov commented on packages/forklift-console-plugin/src/modules/Overview/views/overview/tabs/Details/cards/ChartsCard.tsx:
```diff
@@ -0,0 +1,196 @@
+  value: number;
+}
+
+const toStarted = (m: V1beta1Migration): string => m.status.started;
+const toFinished = (m: V1beta1Migration): string => m.status.completed;
```
this two methods can be one that return started time if not finished and finish time if finised (success or fail)

then we can remove the need to give the toDataPoint a method to get the timestamp

rszwajko replied:
The method are used also for differentiating finished migrations from others: a migration without valid `completed` timestamp will be filtered out at next step in the pipeline.

yaacov commented on packages/forklift-console-plugin/src/modules/Overview/views/overview/tabs/Details/cards/ChartsCard.tsx:
```diff
@@ -0,0 +1,196 @@
+const isLast24H = (date: DateTime) => date.diffNow('hours').get('hours') <= 24;
+const toDayLabel = (date: DateTime): string => date.toFormat('LLL dd');
+const toHourLabel = (date: DateTime): string => date.toFormat('HH:mm');
+const createTimeBuckets = (isDaysViewSelected: boolean) =>
```
this should get a timegap in seconds and not be restricted  to days or hours

rszwajko replied:
It's a good practice to leave time calculations to the library. Different time units require different math - check https://moment.github.io/luxon/#/math

yaacov replied:
I'm missing something, seconds are seconds in all time systems and units.

using  base units (seconds) is the best and safest way to reduce a list into buckets (in this specific case using the base unit seconds is safer then using a library), please use seconds for this calculation.

yaacov commented:
LGTM

I have some issues with:
a. making the bucketing system more genral
b. making the cards layout work on large screens.

I will address this issues in a followup, great job :fire:
